store/postgres/src/fork.rs

```diff
@@ -140,49 +144,79 @@ impl SubgraphFork {
                     _ => None,
                 });
 
-        if let None = entity {
+        if entity.is_none() {
             return Err(StoreError::ForkFailure(format!(
-                "Unexpected error during query inference! No object type definition with entity type `{}` found in the GraphQL schema supplied by the user.",
+                "No object type definition with entity type `{}` found in the GraphQL schema supplied by the user.",
                 entity_type
             )));
         }
 
-        let fields = &entity.unwrap().fields;
-        let names: Vec<&str> = fields.iter().map(|f| f.name.as_str()).collect();
-
-        let query = Query {
-            query: SubgraphFork::query_string(&entity_type.to_lowercase(), &names),
-            variables: Variables { id },
-        };
-        return Ok((query, fields));
+        Ok(&entity.unwrap().fields)
     }
 
-    fn query_string(entity_type: &str, fields: &[&str]) -> String {
-        format!(
+    fn query_string(&self, entity_type: &str, fields: &[Field]) -> Result<String, StoreError> {
+        let names = fields
+            .iter()
+            .map(|f| {
+                let fname = f.name.to_string();
+                let ftype = f.field_type.to_string().replace(&['!', '[', ']'], "");
+                match ValueType::from_str(&ftype) {
+                    Ok(_) => fname,
+                    Err(_) => {
+                        format!("{} {{ id }}", fname,)
+                    }
+                }
+            })
+            .collect::<Vec<String>>();
+
+        Ok(format!(
             "\
 query Query ($id: String) {{
     {}(id: $id, subgraphError: allow) {{
         {}
     }}
 }}",
-            entity_type,
-            fields.join("\n        ").trim(),
-        )
+            entity_type.to_camel_case(),
+            names.join(" ").trim(),
+        ))
     }
 
     fn extract_entity(
         raw_json: &str,
         entity_type: &str,
         fields: &Vec<Field>,
-    ) -> Result<Entity, StoreError> {
+    ) -> Result<Option<Entity>, StoreError> {
         let json: serde_json::Value = serde_json::from_str(raw_json).unwrap();
-        let json = &json["data"][entity_type.to_lowercase()];
+        let entity = &json["data"][entity_type.to_lowercase()];
+
+        if entity.is_null() {
+            return Ok(None);
+        }
 
         let map: HashMap<Attribute, Value> = {
             let mut map = HashMap::new();
             for f in fields {
-                let value = json.get(&f.name).unwrap().clone();
-                let value = RValue::from(value);
+                if f.is_derived() {
+                    // Derived fields cannot be store.set'ed.
```

Comment:
I'm confused about this comment, can you clarify it for me please?

I know that derived fields cannot be `store.set`'ed, but doesn't the debug fork fetching only happens when **getting** data?

I'm probably missing something, and it might be the case to actually return an error here? (like [this one](https://github.com/graphprotocol/graph-node/blob/399ee444d1023a95f82404d14bb5e4534d827678/graph/src/data/store/mod.rs#L677) we already have)

Reply:
The fetched data must be stored locally, so that the local store is synced once and moves on, rather than constantly fetching new data from the endpoint, which at some point will be outdated and cause unexpected behavior.

That's why fetched entities need to also be `store.set`'ed, but the derived fields can be ignored, since they are never `store.set`'ed.

Now to think of it, I believe I should nevertheless return the derived fields as part of the entity - not completely ignore them. `store.get` resolves derived fields, right?

Reply:
Oh that makes sense. Hmm about the resolution of derived fields, I can't recall if they show up or not on a `store.get`.

One way to be sure is to add a simple dummy `log` to a regular subgraph to be sure if they are resolved or not.

Reply:
Ok, I've played around with [CryptoPunks](https://github.com/itsjerryokolo/CryptoPunks). `dbg!` printing all entities resulting from `store.get`, no one contains derived fields. Thus, I think we can conclude that store.get does not resolve derived fields.

So, we can safely ignore the derived fields fetched from the endpoint. It seems like derived fields resolutions happens at the GraphQL endpoint level, rather than WASM runtime level.

Please correct me if something is wrong in my reasoning.